# **TASK 01 : Translator**

##You need to translate each word or sentence from English to Spanish, French and German 



In [ ]:
!pip install deep-translator

In [2]:
import pandas as pd
import numpy as np
import re
from deep_translator import GoogleTranslator

In [3]:
df = pd.read_csv("/content/English.csv")
df.head()

English words/sentences
0                     Hi.
1                    Run!
2                    Run!
3                    Who?
4                    Wow!

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 175621 entries, 0 to 175620
Data columns (total 1 columns):
 #   Column                   Non-Null Count   Dtype 
---  ------                   --------------   ----- 
 0   English words/sentences  175621 non-null  object
dtypes: object(1)
memory usage: 1.3+ MB


In [5]:
df.describe()

English words/sentences
count                                              175621
unique                                             123100
top     I can't tell you how happy I am that you've co...
freq                                                   32

* We can clear see that the dataset has nearly 1,23,100 unique data out of 175621.  
* Progess that much Data will take More compile time.  
* so I am going to take 1000 data randomly and do the Process.. 

In [19]:
df1 = df.drop_duplicates()
df1.head()

English words/sentences
0                     Hi.
1                    Run!
3                    Who?
4                    Wow!
5                   Fire!

In [20]:
df1['clean_words'] = [re.sub(r'[^\w\s]', '', i) for i in list(df1['English words/sentences'])]
df1.head()

<ipython-input-20-786bf0d229f9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['clean_words'] = [re.sub(r'[^\w\s]', '', i) for i in list(df1['English words/sentences'])]


English words/sentences clean_words
0                     Hi.          Hi
1                    Run!         Run
3                    Who?         Who
4                    Wow!         Wow
5                   Fire!        Fire

In [26]:
# shuffle the DataFrame rows
df1 = df1.sample(frac = 1)
df1 = df1.head(1000)
df1.head()

English words/sentences  \
81924                        He wouldn't even look at me.   
137298             I love Nara, particularly in the fall.   
163228  I made an appointment to see him at seven o'cl...   
45338                              Why don't you shut up?   
16581                                   Tom hates school.   

                                             clean_words  
81924                         He wouldnt even look at me  
137298              I love Nara particularly in the fall  
163228  I made an appointment to see him at seven oclock  
45338                               Why dont you shut up  
16581                                   Tom hates school

In [27]:
df1['To_french'] = df1['clean_words'].apply(lambda x : GoogleTranslator(source='auto', target='fr').translate(x))
df1.head()

English words/sentences  \
81924                        He wouldn't even look at me.   
137298             I love Nara, particularly in the fall.   
163228  I made an appointment to see him at seven o'cl...   
45338                              Why don't you shut up?   
16581                                   Tom hates school.   

                                             clean_words  \
81924                         He wouldnt even look at me   
137298              I love Nara particularly in the fall   
163228  I made an appointment to see him at seven oclock   
45338                               Why dont you shut up   
16581                                   Tom hates school   

                                               To_french  
81924                 Il ne voulait même pas me regarder  
137298          J'aime Nara particulièrement à l'automne  
163228  J'ai pris rendez-vous pour le voir à sept heures  
45338                         Pourquoi ne te tais-tu pas  
16581                                Tom déteste l'école

In [30]:
df1['To_spanish'] = df1['clean_words'].apply(lambda x : GoogleTranslator(source='auto', target='es').translate(x))
df1.head()

English words/sentences  \
81924                        He wouldn't even look at me.   
137298             I love Nara, particularly in the fall.   
163228  I made an appointment to see him at seven o'cl...   
45338                              Why don't you shut up?   
16581                                   Tom hates school.   

                                             clean_words  \
81924                         He wouldnt even look at me   
137298              I love Nara particularly in the fall   
163228  I made an appointment to see him at seven oclock   
45338                               Why dont you shut up   
16581                                   Tom hates school   

                                               To_french  \
81924                 Il ne voulait même pas me regarder   
137298          J'aime Nara particulièrement à l'automne   
163228  J'ai pris rendez-vous pour le voir à sept heures   
45338                         Pourquoi ne te tais-tu pas   
16581                                Tom déteste l'école   

                                            To_spanish  
81924                           Él ni siquiera me miró  
137298         Amo a Nara particularmente en el otoño.  
163228  Hice una cita para verlo a las siete en punto.  
45338                           ¿Por qué no te callas?  
16581                              Tom odia la escuela

In [31]:
df1['To_german'] = df1['clean_words'].apply(lambda x : GoogleTranslator(source='auto', target='de').translate(x))
df1.head()

English words/sentences  \
81924                        He wouldn't even look at me.   
137298             I love Nara, particularly in the fall.   
163228  I made an appointment to see him at seven o'cl...   
45338                              Why don't you shut up?   
16581                                   Tom hates school.   

                                             clean_words  \
81924                         He wouldnt even look at me   
137298              I love Nara particularly in the fall   
163228  I made an appointment to see him at seven oclock   
45338                               Why dont you shut up   
16581                                   Tom hates school   

                                               To_french  \
81924                 Il ne voulait même pas me regarder   
137298          J'aime Nara particulièrement à l'automne   
163228  J'ai pris rendez-vous pour le voir à sept heures   
45338                         Pourquoi ne te tais-tu pas   
16581                                Tom déteste l'école   

                                            To_spanish  \
81924                           Él ni siquiera me miró   
137298         Amo a Nara particularmente en el otoño.   
163228  Hice una cita para verlo a las siete en punto.   
45338                           ¿Por qué no te callas?   
16581                              Tom odia la escuela   

                                         To_german  
81924           Er würde mich nicht einmal ansehen  
137298          Ich liebe Nara besonders im Herbst  
163228  Ich verabredete mich um sieben Uhr mit ihm  
45338                    Warum schweigst du nicht?  
16581                         Tom hasst die Schule

# **TASK 02 : Spell Checker**

## Create a program that needs to automatically correct that spelling from the word or a given sentence. (Language : English)

**Note:  show 100 data point outputs both in words and sentence.**

In [32]:
import numpy as np
import pandas as pd

In [33]:
df = pd.read_fwf("/content/spell-testset2.txt")

In [34]:
df.head()

appeal: apeal
0            employees: emploies
1            encourage: encorage
2           permanent: perminant
3  mathematically: mathematicaly
4                    data: dsata

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 362 entries, 0 to 361
Data columns (total 1 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   appeal: apeal  362 non-null    object
dtypes: object(1)
memory usage: 3.0+ KB


In [36]:
df.describe()

appeal: apeal
count                   362
unique                  362
top     employees: emploies
freq                      1

In [37]:
#splitting the appeal: apeal	 column into Correct_word and Misspelled_word
df[["Correct_word", "Misspelled_word"]]  = df['appeal: apeal'].str.split(" ",n=1, expand = True)
df.head()

appeal: apeal     Correct_word Misspelled_word
0            employees: emploies       employees:        emploies
1            encourage: encorage       encourage:        encorage
2           permanent: perminant       permanent:       perminant
3  mathematically: mathematicaly  mathematically:   mathematicaly
4                    data: dsata            data:           dsata

In [38]:
from textblob import TextBlob

In [39]:
df["After_Correction"]=df['Misspelled_word'].apply(lambda x: str(TextBlob(x.split()[0]).correct()))

In [40]:
df.head(15)

appeal: apeal     Correct_word Misspelled_word  \
0             employees: emploies       employees:        emploies   
1             encourage: encorage       encourage:        encorage   
2            permanent: perminant       permanent:       perminant   
3   mathematically: mathematicaly  mathematically:   mathematicaly   
4                     data: dsata            data:           dsata   
5        permanently: perminantly     permanently:     perminantly   
6          hierarchal: hierachial      hierarchal:      hierachial   
7               proviso: provisoe         proviso:        provisoe   
8                 moving: moveing          moving:         moveing   
9                     allow: alow           allow:            alow   
10                credit: creadit          credit:         creadit   
11            available: availble       available:        availble   
12    traditionally: traditionaly   traditionally:    traditionaly   
13           adaptable: adabtable       adaptable:       adabtable   
14                  later: latter           later:          latter   

   After_Correction  
0          employed  
1         encourage  
2         permanent  
3      mathematical  
4              data  
5       permanently  
6        hierachial  
7           proviso  
8            moving  
9               low  
10           credit  
11        available  
12      traditional  
13        adabtable  
14           latter

In [42]:
s = [1 for i,j in zip(df['Correct_word'],df['After_Correction']) if i[:len(i)-1] == j]
print("out of 362, "+str(sum(s))+" were predicted exactly to the original word")

out of 362, 241 were predicted exactly to the original word


* I have tried few different library to attain the exact word of the mispelled word.  
* But the result was same every time ....  
* Only 60 percent of the mispelled words were corrected exactly to the original word....while others were corrected which is nearly close to the mispelled word.  
* For Eg:  
  * original word    : employees  
  * mispelled word   : emploies  
  * After correction : employed  